### PR TITLE
NanoProd for CMSSW_14_0

### PR DESCRIPTION
**Pull request by mbluj**
This PR is to update NanoProd to 14_0 releases (NanoAOD v13/v14) and to add lepton time-life info.
Changes introduced by this PR are as follows:
* change of default CMSSW to 14_0_1 (current production version of 14_0 series) along with a few other small changes in `env.sh` script,
* removal of `NanoProd/plugins/PATTauSignalCandidatesProducer.cc` which is part of official CMSSW since 13_3.
* adjusted `NanoProd/python/customize.py` to 14_0 nanoAOD workflow and to actually add lepton time-life info which not part of the official nanoAOD workflow.

The updated setup was tested locally with TTbar samples: standard TTSemiLeptonis 2018 and 2022 relval ones.

Note: removal of `NanoProd/plugins/VertexBSTableProducer.cc` module can be considered as related info is added with standard CMSSW module.

This PR supersedes https://github.com/cms-tau-pog/NanoProd/pull/32 which will be closed.

### CODE REVIEW

File: NanoProd/python/customize.py [hunk 1]
```diff
@@ -60,50 +60,50 @@ def customizeTaus(process):
   process.tauTable.variables.leadTkNormChi2 = Var("leadingTrackNormChi2()", float, doc="normalized chi2 of the leading track", precision=10)
   process.tauTable.variables.leadChCandEtaAtEcalEntrance = Var("etaAtEcalEntranceLeadChargedCand", float, doc="eta of the leading charged candidate at the entrance of the ECAL", precision=10)
 
-  process.tauSignalCands = cms.EDProducer("PATTauSignalCandidatesProducer",
-    src = process.tauTable.src,
-    storeLostTracks = cms.bool(True)
+  #for backward compatibility (full set of candVars with redundant charge and mass not stored in official NanoAOD)
```
Comment (kandrosov) [inline NanoProd/python/customize.py]: I think this backward compatibility is not necessary. I suggest removing it.

File: env.sh [hunk 2]
```diff
@@ -75,19 +86,25 @@ action() {
 
   local os_version=$(cat /etc/os-release | grep VERSION_ID | sed -E 's/VERSION_ID="([0-9]+).*"/\1/')
   if [[ $os_version < 8 ]] ; then
-    local os_prefix="cc"
+    local os_prefix="slc"
```
Comment (kandrosov) [inline env.sh]: there is no `cmssw-slc7`. I suggest implementing function:
```sh
get_os_prefix() {
  local os_version=$1
  local for_global_tag=$2
  if (( $os_version >= 8 )); then
    echo el
  elif (( $os_version < 6 )); then
    echo error
  else
    if [[ $for_global_tag == 1 || $os_version == 6 ]]; then
      echo slc
    else
      echo cc
    fi
  fi
}
```
And then:
```sh
local os_prefix=$(get_os_prefix $os_version)
```

File: env.sh [hunk 3]
```diff
@@ -75,19 +86,25 @@ action() {
...
+  local default_cmssw_ver=CMSSW_14_0_1
+  #local target_os_version=7
   local target_os_version=8
-  local target_os_prefix="el"
+  #local target_os_version=9
+  if [[ $target_os_version < 8 ]] ; then
+    local target_os_prefix="slc"
+  else
+    local target_os_prefix="el"
+  fi
```
Comment (kandrosov) [inline env.sh]:
```suggestion
  local target_os_prefix=$(get_os_prefix $target_os_version)
  local target_os_gt_prefix=$(get_os_prefix $target_os_version 1)
```

File: env.sh [hunk 4]
```diff
@@ -75,19 +86,25 @@ action() {
...
   local target_os=$target_os_prefix$target_os_version
   export DEFAULT_CMSSW_BASE="$ANALYSIS_PATH/soft/$default_cmssw_ver"
 
-  run_cmd install_cmssw el8_amd64_gcc11 $default_cmssw_ver $node_os $target_os
+  run_cmd install_cmssw ${target_os}_amd64_gcc11 $default_cmssw_ver $node_os $target_os
```
Comment (kandrosov) [inline env.sh]:
```suggestion
  run_cmd install_cmssw ${target_os_gt_prefix}${target_os_version}_amd64_gcc11 $default_cmssw_ver $node_os $target_os
```

**Review by kandrosov — CHANGES_REQUESTED** [review]
Thank you @mbluj 

I agree that `NanoProd/plugins/VertexBSTableProducer.cc` can be removed (preferably within this PR)

File: env.sh (hunk 2 above)
Comment (mbluj) [reply env.sh]: right, thanks for suggestion.

**Comment by mbluj** [comment]
@kandrosov PR updated following your suggestions.

**Comment by kandrosov** [comment]
@mbluj thank you!
I'm sorry, I forgot one point: why do we need `install_cmssw_addons`? What is missing in 14_0_1? Could these modifications be propagated to the central CMSSW? If possible, I would prefer to avoid using addons because they make `NanoProd` tau-code-dependent. Currently, all tau (or H->tautau) dependency is concentrated in configs and customization.

**Comment by mbluj** [comment]
The `install_cmssw_addons` function is not needed for 14_0_1. It was needed for 13_3_X and has nontrivial effect only for 13_3_1 release. The function might be needed in future and it is why I decided to keep it. But, I have not strong objections if you think it is better to remove it.

**Comment by kandrosov** [comment]
If there are no current plans to include some addons on top of 14_0_X, I propose to remove this function to keep `env.sh` as simple as possible. We didn't use this addon for HTT skims v1, and we plan to base HTT skims v2 on 14_0_1.

**Review by kandrosov — APPROVED** [review]
All looks good to me. I have created [HTT_skim_v1](https://github.com/cms-tau-pog/NanoProd/tree/HTT_skim_v1) branch that should be used to finalize (or update) HTT skim v1. If there are no objections, I'm going to merge this PR into `main`.

**Pull request merged**